main/src/ca/uwaterloo/flix/language/errors/TypeError.scala

```diff
@@ -48,17 +48,75 @@ object TypeError {
          |
          |${code(loc, "unable to generalize the type scheme.")}
          |
+         |The declared type does not match the inferred type:
+         |
          |  Declared: ${cyan(FormatScheme.formatSchemeWithoutConstraints(declared))}
          |  Inferred: ${magenta(FormatScheme.formatSchemeWithoutConstraints(inferred))}
          |""".stripMargin
     }
 
-    /**
-      * Returns a formatted string with helpful suggestions.
-      */
+    def explain(formatter: Formatter): Option[String] = Some(
+      """The declared type is more polymorphic than the inferred type.
+        |
+        |Typically:
+        |
+        |  (a) the declared type signature is incorrect, or
+        |  (b) the expression body is incorrect.
+        |""".stripMargin)
+  }
+
+  /**
+    * Effect Generalization Error.
+    *
+    * @param declared the declared effect.
+    * @param inferred the inferred effect.
+    * @param loc      the location where the error occurred.
+    */
+  case class EffectGeneralizationError(declared: Type, inferred: Type, loc: SourceLocation) extends TypeError {
+    def summary: String = s"The inferred effect '${FormatEff.formatEff(inferred)}' cannot be generalized to '${FormatEff.formatEff(declared)}'."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.format)}
+         |>> The inferred effect: '${red(FormatEff.formatEff(inferred))}' cannot be generalized to '${red(FormatEff.formatEff(declared))}'.
+         |
+         |${code(loc, "unable to generalize the effect.")}
+         |""".stripMargin
+    }
+
     def explain(formatter: Formatter): Option[String] = None
   }
 
+  /**
+    * Impure function declared as pure.
+    *
+    * @param declared the declared effect.
+    * @param inferred the inferred effect.
+    * @param loc      the location where the error occurred.
+    */
+  case class ImpureDeclaredAsPure(declared: Type, inferred: Type, loc: SourceLocation) extends TypeError {
+    def summary: String = "Non-pure function declared as pure. A function with side-effects must be declared as impure."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.format)}
+         |>> Non-pure function declared as pure. A function with side-effects must be declared as impure.
```

Comment:
Why not just call it `impure` instead of `non-pure`?

Reply:
Hmm. Partly because its not a question of pure or impure .. there is also effect polymorphic. So non-pure means not pure (i.e. impure or effect polymorphic). But yeah, not clear. What to do hmm.

Reply:
Fixed.